### PR TITLE
Added operation mode change possibility to use FireWire mode B

### PR DESCRIPTION
**Pull request by Nurgak**
Allows to change FireWire operation mode to mode B. Is [dependent on this pull](https://github.com/rock-drivers/drivers-camera_interface/pull/1) for `camera_interface` as `OperationMode` was not previously defined there.

### CODE REVIEW

**Comment by jmachowinski** [comment]
looks good

**Comment by pagnamen** [comment]
We have new commits for the following reasons:
1) Applying modes before values so that the values are properly configured in the first run if a mode is changed from auto to manual
2) Changing the name of the SHUTTER functionality (which was wrongly called exposure). Now EXPOSURE manages the exposure and SHUTTER the shutter.
3) (Together with the previous 2 reasons) Being able to change mode, value and power (functionality on/off) for the following DC1394 features: exposure, gain, shutter, saturation, sharpness, gamma, white balance.

This pull request depends on the following pull requests:
a) camera_interface
b) camera_base (orogen)

**Comment by doudou** [comment]
The main problem with this is indeed the change of semantic for exposure... It will break other people's systems that rely on exposure to actually set the shutter.

On the other hand, you guys are probably the only rock-users that have FW cameras nowadays.

I don't see a good way to have a smooth "migration". Would be good to at least warn of that change on rock-users.

**Comment by pagnamen** [comment]
I agree that it is hard to get a smooth transition.
On the other hand, we lost several days of work because the driver was wrong and we were thinking that was the camera's fault.

Before, somebody setting exposure manual would actually have set the shutter to manual. Exposure would have continued to work in auto mode. 
Then, setting an exposure value would have set the shutter time (but only on the code second run because of the wrong order).
But exposure is still auto, so the gain is changed automatically with unpredictable behavior, making it impossible to achieve the desired exposure. The gain was impossible to set because there was no call in `camera_base` to change the gain mode which would always stay in auto.

TL;DR, for those leaving exposure to auto nothing changes (because auto is still the default in orogen). For those setting exposure to manual this will lead to behavior changes but anyway it was impossible to get a logic behavior before, so I doubt anybody was even using it.

I agree if you decide to accept the pull request, we should circulate a message in the rock-users mailing list.

**Comment by doudou** [comment]
I was actually wondering about shutter vs. exposure.

From what you just described, in dc1394, "exposure" controls the gain while "shutter" controls the exposure time (indirectly, since the actual exposure time depends also on the trigger mode). Is that right ?

**Comment by pagnamen** [comment]
Exposure controls both gain and shutter to achieve the desired (by the internal camera logic) image properties if they are all set to auto. If shutter or gain are manual, then exposure controls only the other one. If both shutter and gain are manual, exposure has no effect.
If exposure is manual but on, then it will still influence gain and shutter (if those are auto as well) but the exposure coefficients will not change dynamically (the exposure will though).
Shutter is the only physical parameter (how much time the sensor is counting photons). If you change trigger mode, then the shutter time might not be used at all (but just start on rising edge and stop on falling edge for example or start on rising edge but stop after the set shutter time).
For reference, we are using Grasshopper 2, Bumblebee 2 and Bumblebee 3 cameras from point grey only.

I also found this:
"In IIDC, 'features' refer to a number of image parameters that can be tuned like exposure, white balance, etc... Features are adjusted with a number of specific functions. The name of each feature is relatively easy to understand. The only trick is that the EXPOSURE feature is actually an AUTO exposure function. Its behaviour depends on the manufacturer. "
Here: https://damien.douxchamps.net/ieee1394/libdc1394/api/types/

**Comment by doudou** [comment]
Thanks for the long explanations !

I don't have further comments ... Given that things were actually not working before, this is definitely a step up.

**Comment by m0rsch1** [comment]
Dear Maintainers,

we are looking forward to get these pull requests merged, so we can use the master branch again.
Do we have to do anything else? What are the obstacles to merge the pull requests?

BR,

Moritz@ESA

**Comment by jmachowinski** [comment]
Hi Moritz, 
out camera maintainer is stepping down, as you are pretty
active right now, do you want to take over the maintenance of
these packages ?

**Comment by m0rsch1** [comment]
Hi Janosch,

I could try :) Who has been the old maintainer btw.? And who is able to give me the necessary rights?

BR,

Moritz

**Comment by 2maz** [comment]
@m0rsch1: thx for trying - I sent you a github invite for camera packages and the base interface.

**Comment by doudou** [comment]
> I could try :)

Thanks, and welcome :+1:
